R/reactive-cached.R

```diff
@@ -0,0 +1,236 @@
+#' retrieved from the cache). Anything that consumes the `cachedReactive` must
+#' therefore expect it to return a promise.
+#'
+#' @inheritParams reactive
+#' @param cacheKeyExpr An expression that returns a value that will be hashed
+#'   and used as a cache key. This key should be a unique identifier for the
+#'   value: the assumption is that if the cache key is the same, then the value
+#'   of `expr` is the same. If two separate `cachedReactive`s have the same key,
+#'   the value is assumed to be the same. To avoid this, you can add an
+#'   arbitrary identifier (like an ID string) to the cache key.
+#' @param valueExpr The expression that produces the return value of the
+#'   `cachedReactive`. It will be executed within an [isolate()] scope.
+#' @param cache The scope of the cache, or a cache object. This can be `"app"`
+#'   (the default), `"session"`, or a cache object like a [diskCache()]. See the
+#'   Cache Scoping section for more information.
+#' @param ... Other arguments are ignored.
+#'
+#' @examples
+#' ## Only run examples in interactive R sessions
+#' if (interactive()) {
+#'
+#' # Basic example
+#' shinyApp(
+#'   ui = fluidPage(
+#'     sliderInput("x", "x", 1, 10, 5),
+#'     sliderInput("y", "y", 1, 10, 5),
+#'     "x * y: ",
+#'     verbatimTextOutput("txt")
+#'   ),
+#'   server = function(input, output) {
+#'     r <- cachedReactive(
+#'       # Use input$x and input$y for the cache key. This is a reactive
+#'       # expression.
+#'       list(input$x, input$y),
+#'       {
+#'         # The value expression is an expensive computation
+#'         message("Doing expensive computation...")
+#'         Sys.sleep(2)
+#'         input$x * input$y
+#'       }
+#'     )
+#'     output$txt <- renderText(r())
+#'   }
+#' )
+#'
+#'
+#' # Demo of cache key collisions and how to avoid them
+#' shinyApp(
+#'   ui = fluidPage(
+#'     numericInput("n", "n", 1),
+#'     "n * 2: ", verbatimTextOutput("txt2"),
+#'     "n * 3: ", verbatimTextOutput("txt3"),
+#'     "n * 4: ", verbatimTextOutput("txt4")
+#'
+#'   ),
+#'   server = function(input, output) {
+#'     r2 <- cachedReactive(input$n, {
+#'       input$n * 2
+#'     })
+#'     output$txt2 <- renderText(r2())
+#'
+#'     # BAD: This uses the same cache key as r2 above, and so it will collide with
+#'     # r2.
+#'     r3 <- cachedReactive(input$n, {
+#'       input$n * 3
+#'     })
+#'     output$txt3 <- renderText(r3())
+#'
+#'     # GOOD: By adding an ID string,this uses a different cache key from r2,
+#'     # and so its keys will not collide with those from r2.
+#'     r4 <- cachedReactive(list(input$n, "times4"), {
+#'       input$n * 4
+#'     })
+#'     output$txt4 <- renderText(r4())
+#'   }
+#' )
+#'
+#' }
+#'
+#' @importFrom digest digest
+#' @importFrom promises promise is.promising
+#' @export
+cachedReactive <- function(
+  cacheKeyExpr,
+  valueExpr,
+  ...,
+  domain = getDefaultReactiveDomain(),
+  cache = "app")
```

Comment:
What about ignoreNULL (possibly ignoreInit too but maybe not)?

Reply:
I can't think of many use cases for `ignoreNULL` here. The most common use for `ignoreNULL` in `eventReactive` is to ignore the initial value of an actionButton. Typically, you'd want to not do some computation until the actionButton is clicked for the first time; when it is clicked, then the computation happens. It would presumably draw on other values, reactive or not, to do the computation (if you just computed on the value of the actionButton, there's not much you can do with that).

In the case of `cachedReactive`, you would rarely ever want to use just the actionButton as the cache key. It's actually more likely that you'd want to take a dependency on the actionButton, but **not** use its value for the cache key. In this example below, `button` triggers a computation on `x` and `y`.

```R
cachedReactive(
  cacheKeyExpr = {
    if (is.null(input$button)) req(FALSE)
    isolate(list(input$x, input$y))
  },
  valueExpr = {
    input$x + input$y
  }
)
```

I'm glad that you brought this up, though, because the example here is a useful one to demonstrate.